File: crates/dotnet/dotnet/ChatXdk/StreamEncryptor.cs
using System;

namespace ChatXdk
{
    /// <summary>
    /// Incremental stream encryptor for large payloads.
    ///
    /// <para>Feed plaintext with <see cref="Push"/>; call <see cref="Finish"/>
    /// once to emit the final frame. Always <see cref="Dispose"/> when finished.</para>
    /// </summary>
    public sealed unsafe class StreamEncryptor : IDisposable
    {
        private StreamEncryptorHandle* _handle;
        private bool _disposed;

        private StreamEncryptor(StreamEncryptorHandle* handle)
        {
            _handle = handle;
        }

        /// <summary>
        /// Finalizer — frees the native handle (which holds live secretstream key
        /// state) even if <see cref="Dispose"/> is never called explicitly.
        /// An object becomes finalizable as soon as its last use passes — even while
        /// one of its methods is still executing native code — so every method that
        /// passes <c>_handle</c> across the P/Invoke boundary calls
        /// <see cref="GC.KeepAlive(object)"/> on <c>this</c> after the call, keeping
        /// this finalizer from freeing the handle mid-call.
        /// </summary>
        ~StreamEncryptor() => Dispose(disposing: false);

        internal static StreamEncryptor Create(byte[] conversationKey)
        {
            var keyB64 = Chat.Utf8Z(Convert.ToBase64String(conversationKey));
            StreamEncryptorHandle* handle;
            fixed (byte* keyPtr = keyB64)
                handle = NativeMethods.chat_xdk_stream_encryptor_new(keyPtr);
            if (handle == null)
                throw new ChatXdkException("Invalid conversation key (expected 32 bytes)");
            return new StreamEncryptor(handle);
        }

        /// <summary>
        /// Encrypt a plaintext chunk, returning ciphertext available so far
        /// (may be empty for small inputs).
        /// </summary>
        public byte[] Push(byte[] plaintext)
        {
            ThrowIfDisposed();
            var ptB64 = Chat.Utf8Z(Convert.ToBase64String(plaintext));
            string resultB64;
            fixed (byte* ptPtr = ptB64)
                resultB64 = Chat.ConsumeResult(NativeMethods.chat_xdk_stream_encryptor_push(_handle, ptPtr));
            GC.KeepAlive(this);
            return Convert.FromBase64String(resultB64);
        }

        /// <summary>Emit the final frame and consume the encryptor.</summary>
        public byte[] Finish()
        {
            ThrowIfDisposed();
            var resultB64 = Chat.ConsumeResult(NativeMethods.chat_xdk_stream_encryptor_finish(_handle));
            GC.KeepAlive(this);
            return Convert.FromBase64String(resultB64);
        }

        private void ThrowIfDisposed()
        {
            if (_disposed) throw new ObjectDisposedException(nameof(StreamEncryptor));
        }

        /// <summary>Release the native handle.</summary>
        public void Dispose()
        {
            Dispose(disposing: true);
            GC.SuppressFinalize(this);
        }

        private void Dispose(bool disposing)
        {
            // Both the finalizer and Dispose() paths end here; only the
            // unmanaged handle needs freeing.
            if (_disposed) return;
            _disposed = true;
            if (_handle != null)
            {
                NativeMethods.chat_xdk_stream_encryptor_free(_handle);
                _handle = null;
            }
        }
    }
}

File: crates/jvm/java/chatxdk/src/main/java/com/x/chatxdk/StreamEncryptor.java
package com.x.chatxdk;

import com.sun.jna.Memory;
import com.sun.jna.Pointer;

import java.lang.ref.Cleaner;
import java.lang.ref.Reference;
import java.util.Base64;

/**
 * Incremental stream encryptor for large payloads.
 *
 * <p>Feed plaintext with {@link #push}; call {@link #finish} once to emit the
 * final frame. Always {@link #close} the instance to release native resources.
 *
 * <p>Not thread-safe: do not share one instance across threads without external
 * synchronization.
 */
public final class StreamEncryptor implements AutoCloseable {

    /**
     * Frees a native stream-encryptor handle (which holds live secretstream key
     * state). Registered with the shared cleaner; the cleaner runs a registered
     * action at most once, so an explicit {@code close()} and the GC backstop
     * can never double-free.
     *
     * <p>An object becomes phantom-reachable as soon as its last use passes —
     * even while one of its methods is still executing native code — so every
     * method that passes {@code handle} to the native library calls
     * {@link Reference#reachabilityFence} on {@code this} in a {@code finally}
     * block, keeping the cleaner from freeing the handle mid-call.
     */
    private static final class FreeHandle implements Runnable {
        private final Pointer handle;

        FreeHandle(Pointer handle) {
            this.handle = handle;
        }

        @Override
        public void run() {
            ChatNative.INSTANCE.chat_xdk_stream_encryptor_free(handle);
        }
    }

    private Pointer handle;
    private boolean closed;
    private final Cleaner.Cleanable cleanable;

    private StreamEncryptor(Pointer handle) {
        this.handle = handle;
        this.cleanable = Chat.CLEANER.register(this, new FreeHandle(handle));
    }

    static StreamEncryptor create(byte[] conversationKey) {
        try (Memory key = FfiStrings.utf8(Base64.getEncoder().encodeToString(conversationKey))) {
            Pointer h = ChatNative.INSTANCE.chat_xdk_stream_encryptor_new(key);
            if (h == null) {
                throw new ChatXdkException("Invalid conversation key (expected 32 bytes)");
            }
            return new StreamEncryptor(h);
        }
    }

    /**
     * Encrypt a plaintext chunk, returning ciphertext available so far (may be
     * empty for small inputs).
     */
    public byte[] push(byte[] plaintext) {
        throwIfClosed();
        try (Memory pt = FfiStrings.utf8(Base64.getEncoder().encodeToString(plaintext))) {
            String resultB64 =
                    FfiStrings.consume(ChatNative.INSTANCE.chat_xdk_stream_encryptor_push(handle, pt));
            return Base64.getDecoder().decode(resultB64);
        } finally {
            Reference.reachabilityFence(this);
        }
    }

    /** Emit the final frame and consume the encryptor. */
    public byte[] finish() {
        throwIfClosed();
        try {
            String resultB64 =
                    FfiStrings.consume(ChatNative.INSTANCE.chat_xdk_stream_encryptor_finish(handle));
            return Base64.getDecoder().decode(resultB64);
        } finally {
            Reference.reachabilityFence(this);
        }
    }

    private void throwIfClosed() {
        if (closed) {
            throw new IllegalStateException("StreamEncryptor is closed");
        }
    }

    @Override
    public void close() {
        if (!closed) {
            closed = true;
            handle = null;
            cleanable.clean();
        }
    }
}
